File: app/go/mnist/eval_mnist.go
package main

import (
	"fmt"
	"log"
	"math"

	"codeberg.org/sbinet/npyio/npz"
	ort "github.com/yalue/onnxruntime_go"
)

const (
	channels   = int64(1)
	height     = int64(28)
	width      = int64(28)
	numClasses = int64(10)
)

func main() {
	// ====== npz からテストデータ読み込み ======
	r, err := npz.Open("/workspace/app/data/mnist_test_normalized.npz")
	if err != nil {
		log.Fatalf("failed to open npz: %v", err)
	}
	defer r.Close()
	var xFlat []float32 // [N * (1*28*28)]
	if err := r.Read("x.npy", &xFlat); err != nil {
		log.Fatalf("failed to read x from npz: %v", err)
	}

	var y []int64 // [N]
	if err := r.Read("y.npy", &y); err != nil {
		log.Fatalf("failed to read y from npz: %v", err)
	}

	numSamples := len(y)
	if numSamples == 0 {
		log.Fatal("empty test set")
	}
	imgSize := int(channels * height * width)
	if len(xFlat) != numSamples*imgSize {
		log.Fatalf("unexpected x size: got %d, want %d (= %d * %d)",
			len(xFlat), numSamples*imgSize, numSamples, imgSize)
	}

	fmt.Printf("loaded test set: N=%d, xFlat=%d, y=%d\n",
		numSamples, len(xFlat), len(y))

	// ====== ONNX Runtime 初期化 ======
	// libonnxruntime.so の場所は環境に合わせて変更してください
	ort.SetSharedLibraryPath("/usr/local/lib/libonnxruntime.so")

	if err := ort.InitializeEnvironment(); err != nil {
		log.Fatalf("failed to init onnxruntime env: %v", err)
	}
	defer func() {
		if err := ort.DestroyEnvironment(); err != nil {
			log.Printf("failed to destroy env: %v", err)
		}
	}()

	// ====== 1枚ずつ推論するシンプルな形にする ======
	// 入力: [1,1,28,28]
	inputShape := ort.NewShape(1, channels, height, width)
	inputData := make([]float32, imgSize) // 1枚ぶんのバッファ
	inputTensor, err := ort.NewTensor(inputShape, inputData)
	if err != nil {
		log.Fatalf("failed to create input tensor: %v", err)
	}
	defer inputTensor.Destroy()

	// 出力: [1,10]
	outputShape := ort.NewShape(1, numClasses)
	outputTensor, err := ort.NewEmptyTensor[float32](outputShape)
	if err != nil {
		log.Fatalf("failed to create output tensor: %v", err)
	}
	defer outputTensor.Destroy()

	// ====== セッション作成 ======
	session, err := ort.NewAdvancedSession(
		"/workspace/app/data/mnist_cnn.onnx",
		[]string{"input"},  // ← sess.get_inputs()[0].name の値
		[]string{"logits"}, // ← sess.get_outputs()[0].name の値
		[]ort.Value{inputTensor},
		[]ort.Value{outputTensor},
		nil, // SessionOptions（デフォルトでOKなら nil）
	)
	if err != nil {
		log.Fatalf("failed to create session: %v", err)
	}
	defer func() {
		if err := session.Destroy(); err != nil {
			log.Printf("failed to destroy session: %v", err)
		}
	}()

	// ====== 全サンプルで精度評価（1枚ずつ Run） ======
	correct := 0

	for i := 0; i < numSamples; i++ {
		// xFlat から i枚目を inputData にコピー
		start := i * imgSize
		end := start + imgSize
		copy(inputData, xFlat[start:end])

		// ネットワーク1回実行（inputData の中身を読んで outputTensor を更新）
		if err := session.Run(); err != nil {
			log.Fatalf("failed to run session: %v", err)
		}

		// 出力は [1,10] → 長さ10のスライスとして取得
		logits := outputTensor.GetData()
		if len(logits) != int(numClasses) {
			log.Fatalf("unexpected logits len: %d", len(logits))
		}

		// argmax
		bestIdx := 0
		bestVal := float32(math.Inf(-1))
		for c := 0; c < int(numClasses); c++ {
			if logits[c] > bestVal {
				bestVal = logits[c]
				bestIdx = c
			}
		}

		if int64(bestIdx) == y[i] {
			correct++
		}
	}

	acc := float64(correct) / float64(numSamples)
	fmt.Printf("Test accuracy (Go + ONNX, from npz): %.4f\n", acc)
}

File: app/node/eval_mnist.js
// mnist_eval_onnx.mjs

import fs from "node:fs";
import path from "node:path";
import JSZip from "jszip";
import { load } from "npyjs";
import * as ort from "onnxruntime-node";

const channels = 1;
const height = 28;
const width = 28;
const numClasses = 10;

/**
 * .npz を読み込み、内部の .npy を npyjs でパースして { x, y } を返す
 * Go版: xFlat (float32), y (int64) に相当
 */
async function loadNPZ(npzPath) {
  // ====== npz(zip) を読む ======
  const npzData = fs.readFileSync(npzPath);
  const zip = await JSZip.loadAsync(npzData);
  const result = {};

  // ====== zip 内のすべての .npy を読み取る ======
  for (const filename of Object.keys(zip.files)) {
    if (!filename.endsWith(".npy")) continue;

    const file = zip.files[filename];
    const npyBuffer = await file.async("uint8array");

    // npyjs で decode
    const npyArray = await load(npyBuffer);

    // filename → 配列データを格納
    // 例: "x.npy" → result["x"] = { data, shape, dtype }
    const key = path.basename(filename, ".npy");
    result[key] = npyArray;
  }
  if (!result.x || !result.y) {
    throw new Error("npz 内に x.npy / y.npy が見つかりませんでした");
  }

  return {
    x: result.x, // { data: TypedArray, shape: [...], ... }
    y: result.y,
  };
}

/**
 * Go版と同じロジックで ONNX Runtime による精度評価
 */
async function main() {
  const npzPath = "/workspace/app/data/mnist_test_normalized.npz";
  const onnxPath = "/workspace/app/data/mnist_cnn.onnx";

  // ====== npz から x,y を読み込み ======
  const { x, y } = await loadNPZ(npzPath);

  const xFlat = x.data; // 1次元の float32 or float64 など
  const yRaw = y.data;  // おそらく int64 → BigInt64Array

  // y は BigInt の可能性があるので Number に落とす（0〜9なので安全）
  const labels = Array.from(yRaw, (v) => Number(v));

  const numSamples = labels.length;
  if (numSamples === 0) {
    throw new Error("empty test set");
  }

  const imgSize = channels * height * width;

  if (xFlat.length !== numSamples * imgSize) {
    throw new Error(
      `unexpected x size: got ${xFlat.length}, want ${
        numSamples * imgSize
      } (= ${numSamples} * ${imgSize})`
    );
  }

  console.log(
    `loaded test set: N=${numSamples}, xFlat=${xFlat.length}, y=${labels.length}`
  );

  // ====== ONNX Runtime セッション作成 ======
  const session = await ort.InferenceSession.create(onnxPath);

  // 1枚ぶんの入力バッファ [1,1,28,28]
  const inputTensorData = new Float32Array(imgSize);

  let correct = 0;

  for (let i = 0; i < numSamples; i++) {
    // xFlat から i枚目を inputTensorData にコピー
    const start = i * imgSize;
    const end = start + imgSize;

    // TypedArray 同士なので subarray + set でコピー
    inputTensorData.set(xFlat.subarray(start, end));

    // ONNX Runtime の Tensor を作成
    const inputTensor = new ort.Tensor("float32", inputTensorData, [
      1,
      channels,
      height,
      width,
    ]);

    // 入力名 "input" / 出力名 "logits" は Go 版と同じ前提
    const feeds = { input: inputTensor };

    const results = await session.run(feeds);
    const logitsTensor = results.logits;
    if (!logitsTensor) {
      throw new Error("output 'logits' が見つかりません");
    }

    const logits = logitsTensor.data;
    if (logits.length !== numClasses) {
      throw new Error(
        `unexpected logits len: ${logits.length} (expected ${numClasses})`
      );
    }

    // ====== argmax ======
    let bestIdx = 0;
    let bestVal = -Infinity;
    for (let c = 0; c < numClasses; c++) {
      const v = logits[c];
      if (v > bestVal) {
        bestVal = v;
        bestIdx = c;
      }
    }

    if (bestIdx === labels[i]) {
      correct++;
    }

    // たとえば進捗を出したければこんな感じ（不要なら消してOK）
    // if ((i + 1) % 1000 === 0) {
    //   console.log(`processed ${i + 1}/${numSamples}`);
    // }
  }

  const acc = correct / numSamples;
  console.log(`Test accuracy (Node.js + ONNX, from npz): ${acc.toFixed(4)}`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
